oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/sort/ExternalSortByteArray.java

```diff
@@ -0,0 +1,212 @@
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.function.Function;
+
+/**
+ * Variation of ExternalSort that stores the lines read from intermediate files as byte arrays to avoid the conversion
+ * from byte[] to String and then back.
+ */
+public class ExternalSortByteArray {
+    public static <T> void mergeSortedFilesBinary(List<Path> files, BufferedOutputStream fbw, final Comparator<T> cmp,
+                                                  boolean distinct, Compression algorithm,
+                                                  Function<T, byte[]> typeToByteArray, Function<byte[], T> byteArrayToType)
+            throws IOException {
+        ArrayList<BinaryFileBuffer<T>> bfbs = new ArrayList<>();
+        try {
+            for (Path f : files) {
+                InputStream in = algorithm.getInputStream(Files.newInputStream(f));
+                bfbs.add(new BinaryFileBuffer<>(in, byteArrayToType));
+            }
+            mergeBinary(fbw, cmp, distinct, bfbs, typeToByteArray);
+        } finally {
+            for (BinaryFileBuffer<T> buffer : bfbs) {
+                try {
+                    buffer.close();
+                } catch (Exception ignored) {
+                }
+            }
+            for (Path f : files) {
+                Files.deleteIfExists(f);
+            }
+        }
+    }
+
+    private static <T> int mergeBinary(BufferedOutputStream fbw, final Comparator<T> cmp, boolean distinct,
+                                       List<BinaryFileBuffer<T>> buffers, Function<T, byte[]> typeToByteArray)
+            throws IOException {
+        PriorityQueue<BinaryFileBuffer<T>> pq = new PriorityQueue<>(
+                11,
+                (i, j) -> cmp.compare(i.peek(), j.peek())
+        );
+        for (BinaryFileBuffer<T> bfb : buffers) {
+            if (!bfb.empty()) {
+                pq.add(bfb);
+            }
+        }
+        int rowcounter = 0;
+        T lastLine = null;
+        while (!pq.isEmpty()) {
+            BinaryFileBuffer<T> bfb = pq.poll();
+            T r = bfb.pop();
+            // Skip duplicate lines
+            if (!distinct || lastLine == null || cmp.compare(r, lastLine) != 0) {
+                fbw.write(typeToByteArray.apply(r));
+                fbw.write('\n');
+                lastLine = r;
+            }
+            ++rowcounter;
+            if (bfb.empty()) {
+                bfb.fbr.close();
+            } else {
+                pq.add(bfb); // add it back
+            }
+        }
+        return rowcounter;
+    }
+
+    /**
+     * WARNING: Uses '\n' as a line separator, it will not work with other line separators.
+     */
+    private static class BinaryFileBuffer<T> {
+        private final static int BUFFER_SIZE = 64 * 1024;
```

Comment:
I think this should be configurable with a system property. The reason is, if there are many files (say 10'000), then having them open at the same time can result in out-of-memory.

I wonder if 64 KB is the right size; my experience is that 8 KB doesn't slow down merging significantly.

Reply:
Made it configurable with a default of 16KB.
In my tests, I saw some improvements from 8 to 16KB. And at 16KB, if we have 1000 files to merge, that's 16MB of memory, I don't having 16 instead of 8KB should be an issue at all, as anyway, we should always be much below 1000 files to merge.